src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/pre_compute.cpp

```diff
@@ -0,0 +1,182 @@
+            return false;
+        }
+        LOG_INFO("Rope sin_cos nodes found : sin=" << sin_node->get_name() << ", cos=" << cos_node->get_name());
+        this->matched_cos = cos_node->shared_from_this();
+        this->matched_sin = sin_node->shared_from_this();
+
+        return true;
+    };
+}
+
+ov::npuw::patterns::pre_compute::RopeCacheMatcher::RopeCacheMatcher(const uint32_t max_prompt_len) {
+    auto rpe = std::make_shared<RopePatternLLama2>();
+
+    rpe->callback = [=](ov::pass::pattern::Matcher& m) -> bool {
+        auto inv_freq_size = ov::shape_size(rpe->matched_inv_freq->get_shape());
+
+        LOG_INFO("Making sin-cos cache of size: " << max_prompt_len << "x" << inv_freq_size);
+
+        // shapes  that matches max possible position
+        auto cache = makeCosSinCache(max_prompt_len, rpe->matched_inv_freq);
+
+        // Step 1: Define axis (gather along axis 1)
+        auto axis = ov::op::v0::Constant::create(ov::element::i64, ov::Shape{1}, {1});
+
+        // Step 2: Apply Gather for cos and sin
+        auto gather_cos = std::make_shared<ov::op::v8::Gather>(cache[0], rpe->matched_position_ids, axis);
+        auto gather_sin = std::make_shared<ov::op::v8::Gather>(cache[1], rpe->matched_position_ids, axis);
+        LOG_INFO("Created gather op facilitate LUT search: "<< gather_cos->get_name() << ", " << gather_cos->get_shape());
+
+        // Step 2: convert fp16->fp32
+        auto cos_fp32 = std::make_shared<ov::op::v0::Convert>(gather_cos, ov::element::f32);
+        auto sin_fp32 = std::make_shared<ov::op::v0::Convert>(gather_sin, ov::element::f32);
+
+        // Create the squeeze operation required after gather
+        auto squeeze_cos = std::make_shared<ov::op::v0::Squeeze>(cos_fp32, axis);
+        auto squeeze_sin = std::make_shared<ov::op::v0::Squeeze>(sin_fp32, axis);
+
+        LOG_INFO("Created squeeze_cos op to reduce axis=1: "<< squeeze_cos->get_name() << ", " << squeeze_cos->get_shape());
+        LOG_INFO("Created squeeze_sin op to reduce axis=1: "<< squeeze_sin->get_name() << ", " << squeeze_sin->get_shape());
+
+        LOG_INFO("Rope cos detected at: "<< rpe->matched_cos->get_name() << ", replacing by cache node: "
+            << gather_cos->get_name() << ", " << gather_cos->get_shape());
+        LOG_INFO("Rope sin detected at: "<< rpe->matched_sin->get_name() << ", replacing by cache node: "
+            << gather_sin->get_name() << ", " << gather_sin->get_shape());
+
+        // replacing sin with gather op
+        ov::replace_node(rpe->matched_cos, squeeze_cos);
+        ov::replace_node(rpe->matched_sin, squeeze_sin);
+
+        // TODO: concat_1 still reacheable in partitioning - how to avoid that
+        for (size_t i = 0; i < rpe->matched_concat->get_input_size(); ++i) {
+        //     rpe->matched_concat->input(i).replace_source_output(ov::Output<ov::Node>()); // empty output
+        }
```

Comment:
Technically here's the mechanism but I don't understand its semantics
https://docs.openvino.ai/2025/documentation/openvino-extensibility/transformation-api.html#node-elimination

Reply:
thanks, remove target-input solved the problem